pkg/plugins/manifest.go

```diff
@@ -83,51 +97,127 @@ func readPluginManifest(body []byte) (*pluginManifest, error) {
 }
 
 // getPluginSignatureState returns the signature state for a plugin.
-func getPluginSignatureState(log log.Logger, plugin *PluginBase) PluginSignature {
+func getPluginSignatureState(log log.Logger, plugin *PluginBase) (*PluginSignatureState, error) {
 	log.Debug("Getting signature state of plugin", "plugin", plugin.Id, "isBackend", plugin.Backend)
 	manifestPath := filepath.Join(plugin.PluginDir, "MANIFEST.txt")
 
 	byteValue, err := ioutil.ReadFile(manifestPath)
 	if err != nil || len(byteValue) < 10 {
 		log.Debug("Plugin is unsigned", "id", plugin.Id)
-		return PluginSignatureUnsigned
+		return &PluginSignatureState{
+			Status: PluginSignatureUnsigned,
+		}, nil
 	}
 
 	manifest, err := readPluginManifest(byteValue)
 	if err != nil {
 		log.Debug("Plugin signature invalid", "id", plugin.Id)
-		return PluginSignatureInvalid
+		return &PluginSignatureState{
+			Status: PluginSignatureInvalid,
+		}, nil
 	}
 
 	// Make sure the versions all match
 	if manifest.Plugin != plugin.Id || manifest.Version != plugin.Info.Version {
-		return PluginSignatureModified
+		return &PluginSignatureState{
+			Status: PluginSignatureModified,
+		}, nil
+	}
+
+	// Validate that private is running within defined root URLs
+	if manifest.SignatureType == Private {
+		appURL, err := url.Parse(setting.AppUrl)
+		if err != nil {
+			return nil, err
+		}
+
+		foundMatch := false
+		for _, u := range manifest.RootUrls {
+			rootURL, err := url.Parse(u)
+			if err != nil {
+				log.Warn("Could not parse plugin root URL", "plugin", plugin.Id, "rootUrl", rootURL)
+				return nil, err
+			}
+			if rootURL.Scheme == appURL.Scheme &&
+				rootURL.Host == appURL.Host &&
+				rootURL.RequestURI() == appURL.RequestURI() {
+				foundMatch = true
+				break
+			}
+		}
+
+		if !foundMatch {
+			log.Warn("Could not find root URL that matches running application URL", "plugin", plugin.Id, "appUrl", appURL, "rootUrls", manifest.RootUrls)
+			return &PluginSignatureState{
+				Status: PluginSignatureInvalid,
+			}, nil
+		}
+	}
+
+	// Track which plugin files are included in the signature
+	pluginFileInSignature := make(map[string]bool, len(plugin.Files))
```

Comment:
I think what you really want here is to make a map of files listed in the manifest, and later use that map to tell if there are extra files not listed there.